src/core/lib/iomgr/udp_server.cc

```diff
@@ -597,44 +623,67 @@ int grpc_udp_server_add_port(grpc_udp_server* s,
   }
 
   s->handler_factory = handler_factory;
-  /* Treat :: or 0.0.0.0 as a family-agnostic wildcard. */
-  if (grpc_sockaddr_is_wildcard(addr, &port)) {
-    grpc_sockaddr_make_wildcards(port, &wild4, &wild6);
+  for (size_t i = 0; i < num_listeners; ++i) {
+    /* Treat :: or 0.0.0.0 as a family-agnostic wildcard. */
+    if (grpc_sockaddr_is_wildcard(addr, &port)) {
+      grpc_sockaddr_make_wildcards(port, &wild4, &wild6);
+
+      /* Try listening on IPv6 first. */
+      addr = &wild6;
+      // TODO(rjshade): Test and propagate the returned grpc_error*:
+      GRPC_ERROR_UNREF(grpc_create_dualstack_socket_using_factory(
+          s->socket_factory, addr, SOCK_DGRAM, IPPROTO_UDP, &dsmode, &fd));
+      allocated_port1 =
+          add_socket_to_server(s, fd, addr, rcv_buf_size, snd_buf_size);
+      if (fd >= 0 && dsmode == GRPC_DSMODE_DUALSTACK) {
+        if (port == 0) {
+          /* This is the first time to bind to |addr|. If its port is still
+           * wildcard port, update |addr| with the ephermeral port returned by
+           * kernel. Thus |addr| can have a specific port in following
+           * iterations. */
+          grpc_sockaddr_set_port(addr, allocated_port1);
+          port = allocated_port1;
+        } else if (allocated_port1 >= 0) {
+          /* The following sucessfully created socket should have same port as
+           * the first one. */
+          GPR_ASSERT(port == allocated_port1);
+        }
+        continue;
+      }
+
+      /* If we didn't get a dualstack socket, also listen on 0.0.0.0. */
+      if (port == 0 && allocated_port1 > 0) {
+        /* |port| hasn't been assigned to an emphemeral port yet, |wild4| must
+         * have a wildcard port. Update it with the emphemeral port created
+         * during binding.*/
+        grpc_sockaddr_set_port(&wild4, allocated_port1);
+        port = allocated_port1;
+      }
+      addr = &wild4;
+    }
```

Comment:
I am a bit confused with the logic here. If we failed to create a dual_stack socket, then we must have created an IPV6 socket (i.e dsmode == GRPC_DSMODE_IPV6) - in which case, after setting `add = &wild4`, shouldn't you call `grpc_create_dualstack_socket_using_factory` one more time to create an IPV4 family socket ?

Reply:
I understand what the function was originally trying to do: It is creating either one dual-stack socket or two sockets (one IPV6 and IPV4) or just one IPv4 socket.  Now you added functionality to do that `num_listener` number of times. 
It makes the code a bit tricky and detailed comments would help

Reply:
After setting add = &wild4, grpc_create_dualstack_socket_using_factory() is called a few lines below to create a IPV4 socket.

The only difference in the new logic is updating addr with port after binding as early as possible. So that addr can be used in next iteration with a known port. 
I added more comments, hoping it makes more sense now.

Reply:
I initially thought the 'for-loop' ended right after `addr = &wild4` line :) and hence the confusion.
It makes sense now.